src/main/java/org/opensearch/security/privileges/ActionPrivileges.java

```diff
@@ -901,11 +1022,69 @@ PrivilegesEvaluatorResponse providesExplicitPrivilege(
                     }
                 }
             }
-
             return PrivilegesEvaluatorResponse.insufficient(checkTable)
                 .reason("No explicit privileges have been provided for the referenced indices.")
                 .evaluationExceptions(exceptions);
         }
+
+        PrivilegesEvaluatorResponse apiTokenProvidesIndexPrivilege(
+            PermissionBasedPrivilegesEvaluationContext context,
+            IndexResolverReplacer.Resolved resolvedIndices,
+            Set<String> actions,
+            Boolean explicit
+        ) {
+            Permissions permissions = context.getPermissions();
+            List<ApiToken.IndexPermission> indexPermissions = permissions.getIndexPermission();
+
+            for (String concreteIndex : resolvedIndices.getAllIndices()) {
+                boolean indexHasAllPermissions = false;
+
+                // Check each index permission
+                for (ApiToken.IndexPermission indexPermission : indexPermissions) {
+                    // First check if this permission applies to this index
+                    boolean indexMatched = false;
+                    for (String pattern : indexPermission.getIndexPatterns()) {
+                        if (WildcardMatcher.from(pattern).test(concreteIndex)) {
+                            indexMatched = true;
+                            break;
+                        }
+                    }
+
+                    if (!indexMatched) {
+                        continue;
+                    }
+
+                    // Index matched, now check if this permission covers all actions
+                    Set<String> remainingActions = new HashSet<>(actions);
+                    ImmutableSet<String> resolvedIndexPermissions = actionGroups.resolve(indexPermission.getAllowedActions());
+
+                    for (String permission : resolvedIndexPermissions) {
+                        // Skip global wildcard if explicit is true
+                        if (explicit && permission.equals("*")) {
+                            continue;
+                        }
+
+                        WildcardMatcher permissionMatcher = WildcardMatcher.from(permission);
+                        remainingActions.removeIf(action -> permissionMatcher.test(action));
+
+                        if (remainingActions.isEmpty()) {
+                            indexHasAllPermissions = true;
+                            break;
+                        }
+                    }
+
+                    if (indexHasAllPermissions) {
+                        break; // Found a permission that covers all actions for this index
+                    }
+                }
+
+                if (!indexHasAllPermissions) {
+                    return PrivilegesEvaluatorResponse.insufficient("Insufficient permissions for the index" + concreteIndex);
```

Comment:
This error message does not indicate which permission is missing. It also does not mention any other potential indices which also might have missing permissions. Thus, it is not really user friendly.

I would recommend to use the `CheckTable` class as it is also used by the other index privilege evaluation methods.

Reply:
Additionally, the API contract for `hasIndexPrivilege()` which calls this method defines that situations where there are privileges available for sub-sets of indices are reported with a `isPartiallyOk()` response, combined with the set of available indices:

https://github.com/opensearch-project/security/blob/472f7c7979502606ccb0707908ce5f689dde6164/src/main/java/org/opensearch/security/privileges/ActionPrivileges.java#L165-L179

This is necessary for do_not_fail_on_forbidden functionality. Without that behaviour, that functionality won't work correctly for API tokens.